lib/SILOptimizer/Mandatory/TFDeabstraction.cpp

```diff
@@ -2672,16 +2707,55 @@ void TFDeabstractionPass::run() {
   // of the functions would be dead after the first round, but some stragglers
   // remain as in the example above.
   for (auto &fn : *module) {
-    // In dynamic compilation mode, only deabstract accelerator-only functions.
-    if (llvm::TFDynamicCompilation && !isAcceleratorOnly(fn))
+    // Skip functions that are already deabstracted.
+    if (isAcceleratorOnly(fn) || partitionedFunctions.count(&fn) > 0)
       continue;
 
-    // Skip if it is already partitioned, or if it was ignored only because it
-    // operated on tensor values.
-    if (partitionedFunctions.count(&fn) > 0 ||
-        !tfc.shouldBePartitioned(&fn, /*forceTFFunctions=*/true))
-      continue;
-    TFDeabstraction(*this, fn, tfc, constantEvaluator, PM).doIt();
+    // Deabstract the function and force deabstraction of functions that were
+    // ignored in the earlier pass only because it operated on tensor values.
+    deabstract(fn, /*forceTFFunctions*/ true);
+  }
+  return;
+}
+
+} // namespace
+
+namespace {
+  struct TFDeabstractionPass : public SILModuleTransform {
+    /// The entry point to the transformation, runs deabstraction on an entire
+    /// module.
+    void run() override;
+  };
+}  // end anonymous namespace
+
+void TFDeabstractionPass::run() {
+  SILModule *module = getModule();
+  auto &ctx = module->getASTContext();
+
+  // If the TensorFlow module hasn't been imported by the program, don't do
+  // anything.  This avoids impacting compile time for non-TensorFlow using
+  // Swift programs by doing extraneous analysis.
+  auto tfModule = ctx.getLoadedModule(ctx.Id_TensorFlow);
+  if (!tfModule)
+    return;
+
+  // If we are running on the TensorFlow module itself, do not perform
+  // deabstraction.  It contains a lot of code that processes TensorHandle and
+  // other types as host values, and we do not want to force inline all of these
+  // things together.
+  //
+  // TODO: Rework the heuristics in inlineCalls() to be smarter.  In an ideal
+  // world, we would be lazy about inlining, and only inline calls due to actual
+  // inter-op value uses.
+  if (module->getSwiftModule() == tfModule)
+    return;
+
+  TFDeabstractionHelper helper(*this, module);
+  if (PM->getStageName() == "TensorFlow Partitioning") {
```

Comment:
Instead of matching on stage name, it would be cleaner to just have two passes, and insert them at the right part of hte pass pipeline.  The optimizer version of the partitioning pass should be a function pass, and the mandatory version should be a module pass, right?